File: day6_2/src/position.rs
use std::ops::{Add, AddAssign};

use crate::Direction;

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Position {
    x: i32,
    y: i32,
}

#[allow(non_snake_case)]
pub fn Position(x: i32, y: i32) -> Position {
    Position { x, y }
}

impl Add<Direction> for Position {
    type Output = Position;

    fn add(mut self, direction: Direction) -> Self::Output {
        self += direction;
        self
    }
}

impl AddAssign<Direction> for Position {
    fn add_assign(&mut self, direction: Direction) {
        match direction {
            Direction::Up => self.y -= 1,
            Direction::Right => self.x += 1,
            Direction::Down => self.y += 1,
            Direction::Left => self.x -= 1,
        }
    }
}

File: day06_2/py/position.py
from dataclasses import dataclass
from direction import Direction
from copy import copy

@dataclass(frozen=True)
class Position:
    x: int
    y: int

    def __add__(self, direction: Direction) -> "Position":
        x = self.x
        y = self.y

        match direction:
            case Direction.Up: y -= 1
            case Direction.Right: x += 1
            case Direction.Down: y += 1
            case Direction.Left: x -= 1

        return Position(x, y)